File: framework/languages/java/e2e/RegistrationCodec/Client/src/main/java/systems/zlink/e2e/registrationcodec/client/Support/ClientOptions.java
package systems.zlink.e2e.registrationcodec.client.Support;

import java.io.Reader;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Properties;

public record ClientOptions(
    String serverEndpoint,
    String httpEndpoint,
    String codecRequesterHttpEndpoint,
    String invalidServerEndpoint,
    String scenario,
    String buildDir,
    String logDir,
    String invalidServerConfig) {
    public static ClientOptions load(String[] args) {
        if (args.length != 4 || !"--config".equals(args[0]) || args[1].isBlank()
            || !"--scenario".equals(args[2]) || args[3].isBlank()) {
            throw new IllegalArgumentException("Usage: registration-codec-client --config <path> --scenario <selector>");
        }
        Properties values = new Properties();
        try (Reader reader = Files.newBufferedReader(Path.of(args[1]))) {
            values.load(reader);
        } catch (Exception error) {
            throw new IllegalStateException("Could not load RegistrationCodec client config", error);
        }
        return new ClientOptions(
            required(values, "serverEndpoint"),
            required(values, "httpEndpoint"),
            required(values, "codecRequesterHttpEndpoint"),
            required(values, "invalidServerEndpoint"),
            args[3],
            required(values, "buildDir"),
            required(values, "logDir"),
            required(values, "invalidServerConfig"));
    }
    private static String required(Properties values, String name) {
        String value = values.getProperty(name);
        if (value == null || value.isBlank()) throw new IllegalArgumentException(name + " is required");
        return value;
    }
}

File: framework/languages/java/samples/kotlin/ZoneWorld/Client/src/main/kotlin/systems/zlink/samples/kotlin/zoneworld/client/ClientOptions.kt
package systems.zlink.samples.kotlin.zoneworld.client

import java.io.Reader
import java.nio.file.Files
import java.nio.file.Path
import java.util.Properties
data class ClientOptions(
    val gatewayEndpoint: String,
    val opsEndpoint: String,
    val scenario: String,
) {
    companion object {
        fun load(args: Array<String>): ClientOptions {
            require(args.size == 2 && args[0] == "--config" && args[1].isNotBlank()) {
                "Usage: ZoneWorldClient --config <path>"
            }
            val properties = Properties()
            Files.newBufferedReader(Path.of(args[1])).use { reader: Reader -> properties.load(reader) }
            return ClientOptions(
                gatewayEndpoint = required(properties, "sample.gateway-endpoint"),
                opsEndpoint = required(properties, "sample.ops-endpoint"),
                scenario = properties.getProperty("sample.scenario", "full"),
            )
        }

        private fun required(properties: Properties, key: String): String =
            properties.getProperty(key)?.takeIf { it.isNotBlank() }
                ?: error("$key is required")
    }
}
